tools/js-sdk-release-tools/packages/typescript-codegen-breaking-change-detector/src/utils/ast-utils.ts

```diff
@@ -157,3 +162,78 @@ export function findAllRenameAbleDeclarationsInNode(
   findAllRenameAbleDeclarationsInNodeCore(node, scope, service, renameAbleDeclarations, found);
   return renameAbleDeclarations;
 }
+
+export function getCallableEntityReturnTypeNode(node: Node): TypeNode | undefined {
+  switch (node.getKind()) {
+    case SyntaxKind.MethodSignature:
+      return node.asKindOrThrow(SyntaxKind.MethodSignature).getReturnTypeNode();
+    case SyntaxKind.FunctionDeclaration:
+      return node.asKindOrThrow(SyntaxKind.FunctionDeclaration).getReturnTypeNode();
+    case SyntaxKind.PropertySignature:
+      return node.asKindOrThrow(SyntaxKind.PropertySignature).getTypeNode();
+    case SyntaxKind.CallSignature:
+      return node.asKindOrThrow(SyntaxKind.CallSignature).getReturnTypeNode();
+    default:
+      throw new Error(`Unsupported function kind: ${node.getKindName()}`);
+  }
+}
+
+export function getCallableEntityReturnTypeNodeFromSymbol(symbol: Symbol): TypeNode | undefined {
+  const node = symbol.getValueDeclarationOrThrow();
+  return getCallableEntityReturnTypeNode(node);
+}
+
+export function getCallableEntityParameters(node: Node): ParameterDeclaration[] {
+  switch (node.getKind()) {
+    case SyntaxKind.MethodSignature:
+      return node.asKindOrThrow(SyntaxKind.MethodSignature).getParameters();
+    case SyntaxKind.FunctionDeclaration:
+      return node.asKindOrThrow(SyntaxKind.FunctionDeclaration).getParameters();
+    case SyntaxKind.PropertySignature:
+      return getCallableEntityReturnTypeNode(node)?.asKindOrThrow(SyntaxKind.FunctionType).getParameters() ?? [];
+    case SyntaxKind.CallSignature:
+      return node.asKindOrThrow(SyntaxKind.CallSignature)?.getParameters();
+    default:
+      throw new Error(`Unsupported function kind: ${node.getKindName()}`);
+  }
+}
+
+export function getCallableEntityParametersFromSymbol(symbol: Symbol): ParameterDeclaration[] {
+  const node = symbol.getValueDeclarationOrThrow();
+  return getCallableEntityParameters(node);
+}
+
+export function isSameSignature(left: Signature, right: Signature): boolean {
+  if (left.getTypeParameters().length !== right.getTypeParameters().length) return false;
+  if (left.getParameters().length !== right.getParameters().length) return false;
```

Comment:
Just curious, what is getTypeParameters and its difference between getParameters?

Reply:
`interface a{
  <U>(a: U): string
}`

type parameter is the generic type, in this case, `U`